aioxmpp/blocking/service.py

```diff
@@ -0,0 +1,225 @@
+
+       :param blocked_jids: the newly blocked JIDs
+       :type blocklist: :class:`~collections.abc.Set` of :class:`~aioxmpp.JID`
+
+    .. signal:: on_jids_blocked(blocked_jids)
+
+       Fires when JIDs are unblocked.
+
+       :param unblocked_jids: the now unblocked JIDs
+       :type blocklist: :class:`~collections.abc.Set` of :class:`~aioxmpp.JID`
+
+    Coroutine methods:
+
+    .. automethod:: block_jids
+
+    .. automethod:: unblock_jids
+
+    .. automethod:: unblock_all
+    """
+    ORDER_AFTER = [aioxmpp.DiscoClient]
+
+    def __init__(self, client, **kwargs):
+        super().__init__(client, **kwargs)
+        self._blocklist = None
+        self._lock = asyncio.Lock()
+        self._disco = self.dependencies[aioxmpp.DiscoClient]
+
+        self._bse_token = client.before_stream_established.connect(
+            self._get_initial_blocklist
+        )
+
+    on_jids_blocked = callbacks.Signal()
+    on_jids_unblocked = callbacks.Signal()
+    on_initial_blocklist_received = callbacks.Signal()
+
+    @asyncio.coroutine
+    def _check_for_blocking(self):
+        server_info = yield from self._disco.query_info(
+            self.client.local_jid.replace(
+                resource=None,
+                localpart=None,
+            )
+        )
+
+        if namespaces.xep0191 not in server_info.features:
+            raise RuntimeError("server does not support blocklists!")
+
+    @asyncio.coroutine
+    def _get_initial_blocklist(self):
+        yield from self._check_for_blocking()
+
+        if self._blocklist is None:
+            with (yield from self._lock):
+                iq = aioxmpp.IQ(
+                    type_=aioxmpp.IQType.GET,
+                    payload=blocking_xso.BlockList(),
+                )
+                result = yield from self.client.stream.send(iq)
+                self._blocklist = frozenset(result.items)
+            self.on_initial_blocklist_received(self._blocklist)
+
+    @property
+    def blocklist(self):
+        """
+        :class:`~collections.abc.Set` of JIDs blocked by the account.
+        """
+        return self._blocklist
+
+    @asyncio.coroutine
+    def block_jids(self, jids_to_block):
+        """
+        Add the JIDs in the sequence `jids_to_block` to the client's
+        blocklist.
+        """
+        yield from self._check_for_blocking()
+
+        if not jids_to_block:
+            return
+
+        cmd = blocking_xso.BlockCommand()
+        cmd.items[:] = jids_to_block
+        iq = aioxmpp.IQ(
+            type_=aioxmpp.IQType.SET,
+            payload=cmd,
+        )
+        yield from self.client.stream.send(iq)
+
+    @asyncio.coroutine
+    def unblock_jids(self, jids_to_unblock):
+        """
+        Remove the JIDs in the sequence `jids_to_block` from the
+        client's blocklist.
+        """
+        yield from self._check_for_blocking()
+
+        if not jids_to_unblock:
+            return
+
+        cmd = blocking_xso.UnblockCommand()
```

Comment:
I like to have constructors for such XSOs which make the handling easier. Those constructors need tests then.